### PR TITLE
Skip api call

### PR DESCRIPTION
**Pull request by metobom**
Added a feature that lets skipping API calls. Instead of API call it uses `output`  of `preProcessingSpecifications`, `postProcessingSpecifications`, or both. To skip API call, `operation` must be undefined, `fixedOperationParameters` must be an empty array and one of `preProcessingSpecifications` or `postProcessingSpecifications` must be defined and not be an empty array.

### CODE REVIEW

File: packages/airnode-node/src/api/index.ts [hunk 1]
```diff
@@ -287,11 +288,33 @@ export async function callApi(payload: ApiCallPayload): Promise<LogsData<ApiCall
   if (verificationResult) return verificationResult;
 
   const processedPayload = await preProcessApiSpecifications(payload);
-
-  const [logs, response] = await performApiCall(processedPayload);
-  if (isPerformApiCallFailure(response)) {
-    return [logs, response];
+  const ois = payload.config.ois.find((o: OIS) => o.title === payload.aggregatedApiCall.oisTitle) as OIS;
+  const endpoint = ois.endpoints.find((e: Endpoint) => e.name === payload.aggregatedApiCall.endpointName) as Endpoint;
```
Comment (andreogle) [inline packages/airnode-node/src/api/index.ts:292]: Doesn't this achieve the same thing without the explicit cast?

```suggestion
  const endpoint = ois.endpoints.find((e: Endpoint) => e.name === payload.aggregatedApiCall.endpointName)!;
```

File: packages/airnode-node/src/api/index.ts [hunk 2]
```diff
@@ -287,11 +288,33 @@ export async function callApi(payload: ApiCallPayload): Promise<LogsData<ApiCall
...
+
+  if (!endpoint.operation && endpoint.fixedOperationParameters.length === 0) {
```
Comment (andreogle) [inline packages/airnode-node/src/api/index.ts:294]: This is pretty difficult to follow. Maybe it could be refactored/shortened a bit? Also adjusted the message grammer a little bit. Might also be worth adding some information about the specific invalid endpoint for easier debugging (e.g. the name)

```ts
import isEmpty from 'lodash/isEmpty';

if (!endpoint.operation && isEmpty(endpoint.fixedOperationParameters)) {
  if (isEmpty(endpoint.preProcessingSpecifications) && isEmpty(endpoint.postProcessingSpecifications) {
    const message = `Failed to skip API call. Ensure at least one of 'preProcessingSpecifications' or 'postProcessingSpecifications' is defined and is not an empty array.`;
    const log = logger.pend('ERROR', message);
    return [log, { success: false, errorMessage: message }];
  }
  return processSuccessfulApiCall(payload, { data: processedPayload.aggregatedApiCall.parameters });
}

const [logs, response] = await performApiCall(processedPayload);
if (isPerformApiCallFailure(response)) {
  return [logs, response];
}
return processSuccessfulApiCall(payload, response);
```

Comment (andreogle) [reply packages/airnode-node/src/api/index.ts:294]: It's probably also worth adding a comment explaining what's going on here too, since it's not obvious

File: packages/airnode-node/src/api/index.ts [hunk 3]
```diff
@@ -287,11 +288,33 @@ export async function callApi(payload: ApiCallPayload): Promise<LogsData<ApiCall
...
+        [
+          logger.pend(
+            'ERROR',
+            `Failed to skip API call. Make sure one, or both 'preProcessingSpecifications', 'postProcessingSpecifications' exists and not empty array.`
+          ),
+        ],
+        {
+          success: false,
+          errorMessage: `Failed to skip API call. Make sure one, or both 'preProcessingSpecifications', 'postProcessingSpecifications' exists and not empty array.`,
+        } as SkipApiCallError,
```
Comment (andreogle) [inline packages/airnode-node/src/api/index.ts:309]: What's the reason for casting to this specific interface? The function signature specifies that this returns `Promise<LogsData<ApiCallResponse>>` so it's kinda confusing.

File: packages/airnode-node/src/api/index.ts [hunk 4]
```diff
@@ -287,11 +288,33 @@ export async function callApi(payload: ApiCallPayload): Promise<LogsData<ApiCall
...
+      ];
+    }
+    return processSuccessfulApiCall(payload, { data: processedPayload.aggregatedApiCall.parameters });
+  } else {
```
Comment (andreogle) [inline packages/airnode-node/src/api/index.ts:313]: No need for the `else` since the `if` blocks above already return early in all of the paths

File: packages/airnode-node/src/types.ts [hunk 5]
```diff
@@ -246,6 +246,11 @@ export interface ApiCallErrorResponse {
   errorMessage: string;
 }
 
+export interface SkipApiCallError {
```
Comment (andreogle) [inline packages/airnode-node/src/types.ts:249]: This is a duplicate interface of `ApiCallErrorResponse` above

File: packages/airnode-node/src/api/index.test.ts [hunk 6]
```diff
@@ -414,6 +414,146 @@ describe('callApi', () => {
...
+        success: true,
+        data: {
+          encodedValue: createEncodedValue(ethers.BigNumber.from(25 * 2 + 5)),
+          signature:
+            '0x9796e8ba07c517a43b2ca1416b9c975541345c5b7bf73613738ed84216c4573d1656cbc385acb9111b7077fba65418b59cfe3f59ee81109523664bbc25efcfd71b',
+        },
+      });
+    });
+
+    it('fail when both preProcessingSpecifications and postProcessingSpecifications are empty or undefined', async () => {
```
Comment (andreogle) [inline packages/airnode-node/src/api/index.test.ts:528]:
```suggestion
    it('fails when both preProcessingSpecifications and postProcessingSpecifications are empty or undefined', async () => {
```

Comment (andreogle) [reply packages/airnode-node/src/api/index.test.ts:528]: (This test is also only testing empty arrays, while the description also says "undefined")

File: packages/airnode-node/src/api/index.ts [hunk 7]
```diff
@@ -287,11 +288,33 @@ export async function callApi(payload: ApiCallPayload): Promise<LogsData<ApiCall
...
+            `Failed to skip API call. Make sure one, or both 'preProcessingSpecifications', 'postProcessingSpecifications' exists and not empty array.`
+          ),
+        ],
+        {
+          success: false,
+          errorMessage: `Failed to skip API call. Make sure one, or both 'preProcessingSpecifications', 'postProcessingSpecifications' exists and not empty array.`,
+        } as SkipApiCallError,
+      ];
+    }
+    return processSuccessfulApiCall(payload, { data: processedPayload.aggregatedApiCall.parameters });
```
Comment (andreogle) [inline packages/airnode-node/src/api/index.ts:312]: Could we add a comment here too explaining why this returns the aggregatedApiCall parameters as the result?

File: packages/airnode-node/src/api/index.test.ts [hunk 8]
```diff
@@ -414,6 +414,146 @@ describe('callApi', () => {
...
+        success: true,
+        data: {
+          encodedValue: createEncodedValue(ethers.BigNumber.from(25 * 2)),
+          signature:
+            '0x148d65210b201c3ddd6f8e08cfc29032e9f3c781919eeb80b950dbb984ebbbee289105ea8b591ace166d04b65e334801ea0177bb431775db2b4268d55fa4e2a81c',
+        },
+      });
+    });
+
+    it('skip api call with postProcessingSpecifications', async () => {
```
Comment (andreogle) [inline packages/airnode-node/src/api/index.test.ts:454]:
```suggestion
    it('skips the API call with postProcessingSpecifications and no operation', async () => {
```

File: packages/airnode-node/src/api/index.test.ts [hunk 9]
```diff
@@ -414,6 +414,146 @@ describe('callApi', () => {
       );
     });
   });
+  describe('skip api call', () => {
+    const createEncodedValue = (value: ethers.BigNumber, times = 100_000) =>
+      `0x${value.mul(times).toHexString().substring(2).padStart(64, '0')}`;
+
+    it('skip api call with preProcessingSpecifications', async () => {
```
Comment (andreogle) [inline packages/airnode-node/src/api/index.test.ts:421]:
```suggestion
    it('skips the API call with preProcessingSpecifications and no operation', async () => {
```

**Review by andreogle — CHANGES_REQUESTED** [review]

**Review by andreogle — APPROVED** [review]
Looks great 👌🏻 Nothing more from my end

**Review by amarthadan — APPROVED** [review]
LGTM :+1:
